### PR TITLE
DX-326 Intercept Docs Improvements (Subtitle and Syntax Section)

### PR DESCRIPTION
**Pull request by mccataldo**
closes DX-326 DX-300

### CODE REVIEW

**Review by cowboy — CHANGES_REQUESTED** [review]

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:7]: This is my attempt at trying to delineate between a request (inclusive of the response) and just the request. Does that make sense?

Comment (cowboy) [reply content/api/commands/intercept.md:7]: It does, but I'm not sure the distinction is necessary. Is there a concern that the reader won't be able to use contextual clues to differentiate between the "request is the whole thing" and "request object which is just a part of the whole thing"?

Comment (mccataldo) [reply content/api/commands/intercept.md:7]: In this case, I'm playing it safe. There's a chance someone could think that we're just waiting on the request to go out. However, if you believe this will cause more confusion I can reword it.

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:8]: It was like that at one point but I didn't like it. I'm not passionate about either way.

**Review by cowboy — CHANGES_REQUESTED** [review]
I went through every API "commands" page at https://docs.cypress.io/api/table-of-contents and they all follow the same format:

* H1 - command name
* brief description
* alerts like alias names, etc
* H2 - Syntax
* list of function signatures
* H3 - Usage
* Correct Usage
* (examples)
* Incorrect Usage
* (examples)
* H3 - Arguments
* (content)
* H3 - Yields
* (content)
* H2 - Examples
* (etc)

I'd like to see the intercept docs follow the same format.

* Let's tighten up the initial description and make it more consistent with the other docs pages. Eg. "Spy on or change the behavior or contents of network requests and responses." The rest of the content could be rolled into examples.
* I'm not sure it's wise to effectively create separate Syntax and Usage sections for a "Spying only" version of intercept vs a "Response Stubbing and Spying" version of intercept. The intercept docs should follow the same pattern as other docs, with a single Syntax block followed by a single Usage block. The rest of the content could be rolled into examples.
* Does the "All intercepts are automatically cleared before every test." note belong up at the top, after the initial description?

**Comment by mccataldo** [comment]
> I went through every API "commands" page at https://docs.cypress.io/api/table-of-contents and they all follow the same format:
> 
> * H1 - command name
> * brief description
> * alerts like alias names, etc
> * H2 - Syntax
> * list of function signatures
> * H3 - Usage
> * Correct Usage
> * (examples)
> * Incorrect Usage
> * (examples)
> * H3 - Arguments
> * (content)
> * H3 - Yields
> * (content)
> * H2 - Examples
> * (etc)
> 
> I'd like to see the intercept docs follow the same format.
> 
The direction we're going with docs is the way it's structured now, where there's no Syntax section. Everything after the description and before Examples falls under Usage.

**Comment by mccataldo** [comment]
> * Let's tighten up the initial description and make it more consistent with the other docs pages. Eg. "Spy on or change the behavior or contents of network requests and responses." The rest of the content could be rolled into examples.
Yeah, that list was there before and I was told to update it to what it is. It really belongs in the Network Requests guide. I could replace it with a callout suggesting that guide as a prerequisite.

**Comment by mccataldo** [comment]
> * I'm not sure it's wise to effectively create separate Syntax and Usage sections for a "Spying only" version of intercept vs a "Response Stubbing and Spying" version of intercept. The intercept docs should follow the same pattern as other docs, with a single Syntax block followed by a single Usage block. The rest of the content could be rolled into examples.

The reason this isn't consistent is that its design (I would argue) is not consistent. In other words, this command does a lot more than most commands. 

Here's how it would look if it was consistent: 

## Syntax

```js
cy.intercept(url)
cy.intercept(method, url)
cy.intercept(routeMatcher)
cy.intercept(url, staticResponse)
cy.intercept(method, url, staticResponse)
cy.intercept(routeMatcher, staticResponse)
cy.intercept(url, routeMatcher, staticResponse)
cy.intercept(url, routeHandler)
cy.intercept(method, url, routeHandler)
cy.intercept(routeMatcher, routeHandler)
cy.intercept(url, routeMatcher, routeHandler)
```

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
cy.intercept('/users/**')
cy.intercept('GET', '/users*')
cy.intercept({
  method: 'GET',
  url: '/users*',
  hostname: 'localhost',
})
cy.intercept(
  {
    method: 'POST',
    url: '/users*',
  },
  {
    statusCode: 201,
    body: {
      name: 'Peter Pan',
    },
  }
)
cy.intercept('/users*', { hostname: 'localhost' }, (req) => {
  /* do something with request and/or response */
})
```

At the end of the day, I care most about what will result in the best developer experience.

**Comment by mccataldo** [comment]
> * Does the "All intercepts are automatically cleared before every test." note belong up at the top, after the initial description?

Hehe, I got this question already. My answer is still the same: I don't know. It needs to be called out and I don't have a better place to put it.

**Comment by mccataldo** [comment]
@cowboy have another look at the latest changes and resolve the comments when you're done.

File: content/api/commands/intercept.md
Comment (cowboy) [inline content/api/commands/intercept.md:148]: Should there be a `### Spying` example?

File: content/api/commands/intercept.md
Comment (cowboy) [inline content/api/commands/intercept.md:154]: I'm looking at the "Matching url" and "Matching method" examples, and it's unclear what they actually _do._ Do I also need to alias them with `.as()` for them to be useful?

**Review by cowboy — CHANGES_REQUESTED** [review]

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:122]: This is how it was presented initially and I'd like to keep it that way because one table defines the actual response content while the other defines how it will be delivered.

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:148]: Yeah, but as we talked about, that's not in the scope of this PR.

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:154]: Let's revisit this in the next iteration.

File: content/api/commands/intercept.md
Comment (cowboy) [reply content/api/commands/intercept.md]: For consistency with other API docs, instead of mentioning the API method like so:
> `cy.intercept` can be used to capture and even change the behavior or contents of network requests and responses.

Can you do something like this:
> Capture or change the behavior or contents of network requests and responses.

**Review by cowboy — CHANGES_REQUESTED** [review]
Looks good, I just had one final comment. Also, for some reason, all my comments had been marked as "resolved" already. Not sure if that's a Github glitch... but I un-resolved the one that I had that comment on. Thanks!

**Comment by mccataldo** [comment]
@jennifer-shehane I'd really like your input on this. Keep in mind that the scope of this PR is just around the subtitle and Syntax section. I'm working on the rest in future iterations.

**Comment by mccataldo** [comment]
@amirrustam feel free to provide feedback as well if you'd like.

**Review by cowboy — APPROVED** [review]

File: content/api/commands/intercept.md
Comment (mccataldo) [inline content/api/commands/intercept.md:19]:
```suggestion
<Alert type="info">
```

File: content/api/commands/intercept.md
Comment (mccataldo) [inline content/api/commands/intercept.md:4]: here are some comments about the entire file

**Comment by mccataldo** [comment]
@cowboy before merging this I'd like to discuss the changes that we made around adding the signature for the `staticResponse`. I like this a lot, but in cases where users might be using a TS compiler (for TS or just Intellisense), this could create confusion because now our docs aren't consistent with those types. To be clear, we've never really focussed on this but where this is a more complex command, users might be leaning on types. In the real world, this is what I would have done.

**Comment by cowboy** [comment]
> @cowboy before merging this I'd like to discuss the changes that we made around adding the signature for the `staticResponse`. I like this a lot, but in cases where users might be using a TS compiler (for TS or just Intellisense), this could create confusion because now our docs aren't consistent with those types. To be clear, we've never really focused on this but where this is a more complex command, users might be leaning on types. In the real world, this is what I would have done.

Why don't we merge as-is and we can add a follow-up story to look into this further. If the code can be updated to use types that match the docs, that would be awesome. But we'll most likely need to look into this separately.

**Comment by cowboy** [comment]
> Feel free to merge it if you want. I'm just proceeding cautiously here considering this is a relatively new and complex feature. I also don't want to back-pedal from a change like this and cause confusion in the community.

@amirrustam what do you think?

**Comment by jennifer-shehane** [comment]
@mccataldo Looks good. The structure feels so much more readable, like my brain is not jumping around anymore. 

A couple of notes. 

1. The comments are often cut off. I'm not sure if it was originally written this way or that the prettier reformats them (Gleb mentioned this may be happening in the past but I never confirmed). It's be great to split the comments up into shorter line lengths so they're fully readable on a standard size screen without scrolling.

Screen Shot 2021-06-02 at 8 05 07 AM

2. I like the glob matching section and imagine I will send a lot of people to this section. I'd like to see more examples here - but this may be more appropriate for another PR. I often see people matching on the path only then their url not matching because it has query params. I'd like to see this example. 

```js
cy.intercept('/api/roles')
// will match https://example.com/api/roles
// will not match https://example.com/api/roles?country=US

cy.intercept('/api/roles*')
// will match https://example.com/api/roles
// will match https://example.com/api/roles?country=US

// then an example for matching more complex query params
// that I don't know off the top of my head 
// will match https://example.com/api/roles?utm_medium=Nav&utm_campaign=Docs&utm_source=My%App

// will match https://example.com/api/roles?utm_medium=Nav&utm_campaign=Docs&utm_source=https....
```

I hate to be just docs that teach glob matching, but it's a constant issue, where they think cy.intercept is broken and open an issue because it's not finding their url when they haven't written a correct glob matcher.

**Comment by mccataldo** [comment]
> @mccataldo Looks good. The structure feels so much more readable, like my brain is not jumping around anymore.
> 
> A couple of notes.
> 
> 1. The comments are often cut off. I'm not sure if it was originally written this way or that the prettier reformats them (Gleb mentioned this may be happening in the past but I never confirmed). It'd be great to split the comments up into shorter line lengths so they're fully readable on a standard size screen without scrolling.

Yeah, that could be fixed with linting for sure. I created tasks to add rules for code blocks in markdown. In the meantime, I've been manually adding breaks wherever they're needed.

> 2. I like the glob matching section and imagine I will send a lot of people to this section. I'd like to see more examples here - but this may be more appropriate for another PR. I often see people matching on the path only then their url not matching because it has query params. I'd like to see this example.
> 
> ```js
> cy.intercept('/api/roles')
> // will match https://example.com/api/roles
> // will not match https://example.com/api/roles?country=US
> 
> cy.intercept('/api/roles*')
> // will match https://example.com/api/roles
> // will match https://example.com/api/roles?country=US
> 
> // then an example for matching more complex query params
> // that I don't know off the top of my head 
> // will match https://example.com/api/roles?utm_medium=Nav&utm_campaign=Docs&utm_source=My%App
> 
> // will match https://example.com/api/roles?utm_medium=Nav&utm_campaign=Docs&utm_source=https....
> ```
> 
> I hate to be just docs that teach glob matching, but it's a constant issue, where they think cy.intercept is broken and open an issue because it's not finding their url when they haven't written a correct glob matcher.

If it's a constant issue and we lack the content, then that's enough justification for me. I can add examples or at least a link to examples. I could also add content on testing matches when they don't work. I created a task for myself to git 'er done.

**Review by amirrustam — CHANGES_REQUESTED** [review]
This doc is improving with every iteration!

In addition to in-line comments, I also have the following feedback:

- [ ]  Add header title/labels to alert boxes. 
     Right now most of the alert boxes have not contextual title:
     
CleanShot 2021-06-01 at 11 53 38@2x

CleanShot 2021-06-01 at 11 52 31@2x

In the alert box titles, we can say things like "Note", "Tip", or "Consider X." Or some title that's contextually relevant to the info in the box.

- [ ]  Move GraphQL examples into the main examples section. They don't need their own top-level placement.
-- 

@cowboy and @mccataldo It's not clear to me what the confusion would be with the types and documentation of `StaticResponse`. Is the concern that it can be both a `string` and an `object`? Can you please elaborate?

File: content/api/commands/intercept.md
Comment (cowboy) [reply content/api/commands/intercept.md:11]: I think a better approach to point 2 would be to have a callout at the top of the page, like on the [stub](https://docs.cypress.io/api/commands/stub) and [spy](https://docs.cypress.io/api/commands/spy) pages that links to the [Network Requests](https://docs.cypress.io/guides/guides/network-requests) guide.

**Comment by mccataldo** [comment]
> In the alert box titles, we can say things like "Note", "Tip", or "Consider X." Or some title that's contextually relevant to the info in the box.

Good catch. I'll update those callouts and update [Doc Standards](https://www.notion.so/cypressdx/Documentation-Standards-WIP-f9ed5c4e5bdb44aabc4b13814bd693fa)

> * [ ]   Move GraphQL examples into the main examples section. They don't need their own top-level placement.

I can move this to the main examples section, but this is already a huge doc and there's a separate guide for this content. I'll trim down this content to a single example with a brief comment and a link to the dedicated guide.

> @cowboy and @mccataldo It's not clear to me what the confusion would be with the types and documentation of `StaticResponse`. Is the concern that it can be both a `string` and an `object`? Can you please elaborate?

We clarified this offline.

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:11]: > In the alert box titles, we can say things like "Note", "Tip", or "Consider X." Or some title that's contextually relevant to the info in the box.

Good catch. I'll update those callouts and update [Doc Standards](https://www.notion.so/cypressdx/Documentation-Standards-WIP-f9ed5c4e5bdb44aabc4b13814bd693fa)

> * [ ]   Move GraphQL examples into the main examples section. They don't need their own top-level placement.

I can move this to the main examples section, but this is already a huge doc and there's a separate guide for this content. I'll need to trim it down to a single example with a brief comment and a link to the dedicated guide.

> @cowboy and @mccataldo It's not clear to me what the confusion would be with the types and documentation of `StaticResponse`. Is the concern that it can be both a `string` and an `object`? Can you please elaborate?

We clarified this offline. 

> 1. Let's use the "stub" and "spy" terminology as we do everywhere else.

Sounds good. I'll update it. 

> 2. We have to put back the use cases in the intro. I know this is not consistent with all the other commands, but we have to thoughtfully make exceptions where it makes sense.

I agree with @cowboy on this one. However, I want to point out that the Network Requests guide doesn't contain those Intercept use cases so when I removed that content I created [DX-320](https://linear.app/cypress/issue/DX-320/update-guide-network-requests) to address that.

File: content/api/commands/intercept.md
Comment (mccataldo) [reply content/api/commands/intercept.md:60]: Will do

File: content/api/commands/intercept.md
Comment (amirrustam) [reply content/api/commands/intercept.md:11]: Agree with both of your comments above. So we're good to move forward after all these updates.

As a general rule of thumb, we should interlink to other docs whenever appropriate, which is most of the time.

**Comment by mccataldo** [comment]
@amirrustam 
> * [ ]   Add header title/labels to alert boxes.
I like the idea of callout headers but I was unable to find an example in the API docs, only guides.

**Comment by amirrustam** [comment]
@mccataldo an example is not needed. We can just do "Note", "Tip", or "Consider X" or some title that's contextually relevant to the info in the box.

**Comment by mccataldo** [comment]
@amirrustam I think there's a misunderstanding. I'll connect with you directly to clear up the confusion.

**Comment by mccataldo** [comment]
@cowboy @amirrustam this is ready to merge once you give the thumbs up

**Review by amirrustam — APPROVED** [review]
Approving since unimplemented suggestions will come via follow-up PRs.

**Pull request merged**
